src/enzyme_ad/jax/Implementations/StableHLOAutoDiffOpInterfaceImpl.cpp

```diff
@@ -417,6 +417,84 @@ class AutoDiffBroadcastInDimRev
                           MGradientUtilsReverse *gutils) const {}
 };
 
+class AutoDiffDynamicSliceUpdateRev
+    : public ReverseAutoDiffOpInterface::ExternalModel<
+          AutoDiffDynamicSliceUpdateRev, DynamicUpdateSliceOp> {
+public:
+  LogicalResult createReverseModeAdjoint(Operation *orig, OpBuilder &builder,
+                                         MGradientUtilsReverse *gutils,
+                                         SmallVector<Value> caches) const {
+    auto op = cast<DynamicUpdateSliceOp>(orig);
+    auto operand = op.getOperand();
+    auto update = op.getUpdate();
+
+    SmallVector<Value> startIndices;
+
+    Value diffe;
+    if (!gutils->isConstantValue(operand)) {
+
+      for (auto cache : caches) {
+        startIndices.push_back(gutils->popCache(cache, builder));
+      }
+
+      diffe = gutils->diffe(orig->getResult(0), builder);
+      auto operandDiffe = builder.create<DynamicUpdateSliceOp>(
+          orig->getLoc(), diffe,
+          cast<AutoDiffTypeInterface>(update.getType())
+              .createNullValue(builder, orig->getLoc()),
+          startIndices);
+
+      gutils->addToDiffe(operand, operandDiffe, builder);
+    }
```

Comment:
we should check that there's an optimization which changes the pattern of

x = dyn_update_slice %v, 0's
y = old + x

into
q = dynamic_slice old
t = q + v
y = dynamic_update_slice old, t
^ aka not adding 0's when unnecesary

Reply:
shouldn't some form of constant propagation + DCE do this?